hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3336,6 +3338,16 @@ private TableGroup consumeAttributeJoin(
 			registerEntityNameProjectionUsage( sqmJoin, getActualTableGroup( joinedTableGroup, sqmJoin ) );
 		}
 		registerPathAttributeEntityNameUsage( sqmJoin, ownerTableGroup );
+		if ( !sqmJoin.hasTreats() && sqmJoin.getNodeType().getSqmPathType() instanceof EntityDomainType ) {
+			final EntityDomainType<?> entityDomainType = (EntityDomainType<?>) sqmJoin.getNodeType().getSqmPathType();
+			final String entityName = entityDomainType.getHibernateEntityName();
+			final EntityPersister entityDescriptor = domainModel.findEntityDescriptor( entityName );
```

Comment:
To avoid doing a map lookup it might be possible to use `(EntityPersister) joinedTableGroup.getModelPart().getPartMappingType()` instead, which should never be null.

Reply:
I tried this but I'm sure this would work, as we would need to also handle collections and we would need to differentiate between element / index descriptors. Any idea if there's an alternative (other than what I've already done that requires a map lookup)?

Reply:
I guess you'd have to do something like this:

```
final TableGroup elementTableGroup = joinedTableGroup instanceof PluralTableGroup 
  ? ( (PluralTableGroup) joinedTableGroup ).getElementTableGroup()
  joinedTableGroup;
final EntityPersister entityDescriptor = (EntityPersister) realTableGroup.getModelPart().getPartMappingType();
```

Reply:
Couldn't we need to do this for the plural `getIndexTableGroup()` instead of the element though?

Reply:
A join on a `PluralJoin` is always against the element. A join against the index part will be a `PluralPartJoin` AFAIR.

Reply:
I've applied your suggestion, slightly modified it to account for all cases.